accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,185 @@
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator() { ... }
+        IEnumerator IEnumerable.GetEnumerator() { ... }
+    }
+```
+
+## Activity.Tags
+
+As mentioned earlier, Activity has Tags property which returns the list of the activity tags. Tags are a list of `KeyValuePair<string, string>`. Tags are proven to be not enough for OT scenarios as we need to support more value types (numeric, bool, numeric array, and nested tags). Activity today has the method `public Activity AddTag(string key, string? value)`. The proposal here is to add more `AddTag` overload accepting more types and add a new getter property return the list of all typed tags.
+
+```C#
+    public partial class Activity : IDisposable
+    {
+        public void AddTag(string key, int value) { }
+        public void AddTag(string key, double value) { }
+        public void AddTag(string key, bool value) { }
+        public void AddTag(string key, int [] value) { }
+        public void AddTag(string key, double [] value) { }
+        public void AddTag(string key, bool [] value) { } // how interesting this one?
+        public void AddTag(string key, string [] value) { }
+
+        // The following API is pending the approval in OT spec first before we add it in .NET.
+        public void AddTag(string key, ActivityTagsCollection value) { }
+
+        // We cannot change Tags property so we are introducing a new property with a new name
+        public ActivityTagsCollection TypedTags { get; }
+    }
+```
+
+***<u>OpenQuestion</u>***
+
+We need to decide about the behavior of the old property `Activity.Tags` when adding tags of non-string types values. The options are
+
+- Ignore all non-string typed values and return only the string type value tags.
+- Convert the non-string type values to string before returning it.
+
+I am inclining to the first option to avoid handling ToString with more complex types (e.g. arrays or ActivityTagsCollection).
+
+## Proposal Changing already exposed APIs in previous preview releases
+
+As we are introducing the ActivityTagsCollection, we can take advantage of that now and modify some of the exposed APIs to use this collection instead. Also, we are renaming Attributes to Tags for consistency inside BCL APIs.
+
+### ActivityEvent
+
+```C#
+    // Already exposed APIs
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public ActivityEvent(string name, DateTimeOffset timestamp, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>> Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, ActivityTagsCollection? tags)
+        public ActivityEvent(string name, DateTimeOffset timestamp, ActivityTagsCollection? tags)
+        public ActivityTagsCollection? Tags { get; }
```

Comment:
Should we call this Tags or TypedTags ?

Reply:
I prefer Tags here. we used TypedTags in Activity because it had Tags. Also, here, the return type is `ActivityTagsCollection` which simplicity telling the expected  types.

Reply:
So that we can consistently say - TypedTags in Activity is same as Attributes in Span

Reply:
I am not worried about that. In general, moving forward we should message 'Tags' in BCL are the 'Attributes'. Otherwise, we'll stick forever describing the differences between `tags`, `typed tags` and `attributes`.  Note, the API is called AddTag and I don't think we need to rename it to AddTypedTag.

Reply:
so span.AddAttribute -> activity.AddTag. - this is good. Both support string->int/bool/etc. Same for ActivityEvent.AddTag, ActivityLink.AddTag.

For retrieving full attributes the api is: Activity.TypedTags, ActivityEvent.Tags, ActivityLink.Tags -- this is bit inconsistent. In Span API this would be consistent across Span.Attributes, Link.Attributes, Event.Attributes.

Reply:
let me ask, does `Activity.Tags` still considered Attributes or not? my point here is, I don't like using TypedTags, we used it because of the reason we previously exposed have Tags. That means we shouldn't use it in other places. the other idea we can do if you seeing this a concern, instead of having `Activity.TypedTags` we can have `Activity.GetTags()` instead. doing that we'll be consistent all the way.

Reply:
I think Activity.Tags cannot be considered Attributes as its only a sub-set of the full attributes.

My personal feeling is to have consistency, hence .GetTags() everywhere - Activity, Link, Event would give me consistency. Would love to hear from others as well.

Reply:
> I think Activity.Tags cannot be considered Attributes as its only a sub-set of the full attributes.

If it is subset then it is qualified to be `attributes`. The other way around is wrong. `Tags` is `attributes`. attributes not necessarily be `Activity.Tags`. 

> My personal feeling is to have consistency, hence .GetTags() everywhere - Activity, Link, Event would give me consistency. Would love to hear from others as well.

I am open to all suggestions. but I didn't mean to use GetTags everywhere. GetTags will be in Activity only and Tags will be a property in other places. I am not sure why this will be confusing or inconsistent. anyway, I am thinking to schedule a meeting to go through all feedback here and discuss it in person. I appreciate all feedback I am getting from you and others.

Reply:
I do find the inconsistency annoying between Tags and TypedTags though none of the alternatives have seemed that good either. One suggestion that maybe you guys would like:
- Add a concept of activity Attributes, define Tags as the subset of Attributes that are strings, and via docs encourage new code to use Attributes
- In Activity:
  AddTag(string name, string val) { return AddAttribute(name, val); }
  [new API] AddAttribute(string name, object val)
  IEnumerable Tags { return Attributes.Where(a => a is string); }
  [new API] IEnumerable Attributes
- ActivityTagCollection becomes ActivityAttributeCollection
- ActivityEvent.Tags renamed ActivityEvent.Attributes
- ActivityLink.Tags renamed ActivityLink.Attributes
- Any parameter/variable names in the new APIs that reference 'tags' are replaced with 'attributes'


 I take a little solace that all the Tags/TypedTags properties are intended for listeners and the audience of developers who will ever write an ActivityListener is probably small. So at least if things are a little messy not many people will ever have to deal with it.

Reply:
Agree with Attributes - it is simpler name than TypedTags.

If we have an option to introduce a new property for attributes, could we consider having `Set` as well as `Add`?
Everything you said plus:
activity.SetAttribute() -overriding existing key behavior. (This is what is needed for [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-attributes))

The only notable difference now between Span and Activity is Add vs Set behavior for attributes/tags.

Reply:
Although the name `Attributes` was suggested and dismissed before for the sake of consistency with Tags, I don't mind having it back as @noahfalk suggested. 

I don't think we need to have Add and Set for the attributes. The OT defines the behavior of adding Attributes and we'll just follow that. we'll have only SetAttribute. https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-attributes

Reply:
We need AddAttribute() internally to handle AddTag() right? Asuming AddTag will be modified to simply do AddAttribute internally.

But anyway - OT only need SetAttribute, and we are good: :)

Reply:
> We need AddAttribute() internally to handle AddTag() right? Asuming AddTag will be modified to simply do AddAttribute internally.

These are implementation details. I don't think we need to add a new API just to handle the implementation.

Reply:
Yup.